File: 81.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1, mid;
        while (l <= r) {
            mid = (r-l)/2 + l;
            
            if(nums[mid] == target) return true;
            if (nums[l] < nums[mid]) {
                if (nums[l] <= target && target <= nums[mid]) r = mid - 1;
                else l = mid + 1;
            } else if (nums[l] == nums[mid]) {
                l++;
            }
            else {
                if (nums[mid+1] <= target && target <= nums[r]) l = mid + 1;
                else r = mid - 1;
            }
        }
        return false;
    }
};

File: 81.java
class Solution {
    public boolean search(int[] nums, int target) {
        int l = 0, r = nums.length - 1, m;
        while (l <= r) {
            m = (r-l)/2 + l;
            if (nums[m] == target) return true;
            if (nums[m] < nums[r]) {
                if (target > nums[r] || target < nums[m]) r = m - 1;
                else l = m + 1;
            } else if (nums[m] > nums[r]) {
                if (target > nums[m] || target < nums[l]) l = m + 1;
                else r = m - 1;
            } else if (nums[m] == nums[r]) {
                r--;
            }
        }
        return false;
    }
}
